src/pkg/packager/filters/deploy.go

```diff
@@ -0,0 +1,194 @@
+		matchedRequests := map[string]bool{}
+
+		// NOTE: This does not use forIncludedComponents as it takes group, default and required status into account.
+		for _, groupKey := range orderedComponentGroups {
+			var groupDefault *types.ZarfComponent
+			var groupSelected *types.ZarfComponent
+
+			for _, component := range groupedComponents[groupKey] {
+				// Ensure we have a local version of the component to point to (otherwise the pointer might change on us)
+				component := component
+
+				selectState, matchedRequest := includedOrExcluded(component.Name, f.requestedComponents)
+
+				if !isRequired(component, useRequiredLogic) {
+					if selectState == excluded {
+						// If the component was explicitly excluded, record the match and continue
+						matchedRequests[matchedRequest] = true
+						continue
+					} else if selectState == unknown && component.Default && groupDefault == nil {
+						// If the component is default but not included or excluded, remember the default
+						groupDefault = &component
+					}
+				} else {
+					// Force the selectState to included for Required components
+					selectState = included
+				}
+
+				if selectState == included {
+					// If the component was explicitly included, record the match
+					matchedRequests[matchedRequest] = true
+
+					// Then check for already selected groups
+					if groupSelected != nil {
+						return []types.ZarfComponent{}, fmt.Errorf(lang.PkgDeployErrMultipleComponentsSameGroup, groupSelected.Name, component.Name, component.DeprecatedGroup)
+					}
+
+					// Then append to the final list
+					selectedComponents = append(selectedComponents, component)
+					groupSelected = &component
+				}
+			}
+
+			// If nothing was selected from a group, handle the default
+			if groupSelected == nil && groupDefault != nil {
+				selectedComponents = append(selectedComponents, *groupDefault)
+			} else if len(groupedComponents[groupKey]) > 1 && groupSelected == nil && groupDefault == nil {
+				// If no default component was found, give up
+				componentNames := []string{}
+				for _, component := range groupedComponents[groupKey] {
+					componentNames = append(componentNames, component.Name)
+				}
+				return []types.ZarfComponent{}, fmt.Errorf(lang.PkgDeployErrNoDefaultOrSelection, strings.Join(componentNames, ", "))
+			}
+		}
+
+		// Check that we have matched against all requests
+		var err error
+		for _, requestedComponent := range f.requestedComponents {
+			if _, ok := matchedRequests[requestedComponent]; !ok {
+				closeEnough := []string{}
+				for _, c := range pkg.Components {
+					d := levenshtein.ComputeDistance(c.Name, requestedComponent)
+					if d <= 5 {
+						closeEnough = append(closeEnough, c.Name)
+					}
+				}
+				failure := fmt.Errorf(lang.PkgDeployErrNoCompatibleComponentsForSelection, requestedComponent, strings.Join(closeEnough, ", "))
+				if err != nil {
+					err = fmt.Errorf("%w, %w", err, failure)
+				} else {
+					err = failure
+				}
```

Comment:
This could create a kind of crazy error message if it looped a few times.  It might be best to put everything together outside of the loop (a simple flag error boolean and a list of close enoughs would work to then return an error - it would also be a little simpler logic wise).

Reply:
kk, fail on first error here to reduce craziness